src/sphmultipcqueue.cpp

```diff
@@ -0,0 +1,599 @@
+	block_size_t rc = 0;
+
+	if (SOMSASCheckBlockSigAndType ((SASBlockHeader *) headerBlock, SAS_RUNTIME_PCQUEUE_TM)) {
+		sas_read_barrier ();
+		unsigned short stride = headerBlock->default_entry_stride;
+		longPtr_t qlo = headerBlock->startq;
+		longPtr_t qhi = headerBlock->endq;
+		longPtr_t tail = headerBlock->qtail;
+		longPtr_t temp = headerBlock->qhead;
+		do {
+			SPHLFEntryHeader_t *entryPtr = (SPHLFEntryHeader_t *)temp;
+			if (entryPtr->entryID.detail.allocated || entryPtr->entryID.detail.valid) break;
+			rc += stride;
+			temp += stride;
+			if (temp >= qhi) temp = qlo;
+		} while (temp != tail);
+	} else {
+		debug_printf("%s(%p) type check failed\n",__FUNCTION__, queue);
+	}
+	debug_printf("%s=%lu\n",__FUNCTION__,rc);
+	return rc;
+}
+
+sphLFEntryID_t
+SPHMPMCQGetEntryTemplate (SPHMPMCQ_t queue)
+{
+	SPHMPMCQHeader *headerBlock = (SPHMPMCQHeader *) queue;
+	longPtr_t alloc_round = 0;
+	sphLFEntry_t entrytemp;
+
+	entrytemp.idUnit = 0;
+	if (SOMSASCheckBlockSigAndType ((SASBlockHeader *) headerBlock, SAS_RUNTIME_PCQUEUE_TM)) {
+		/* for Strided alloc increment is pre rounded */
+		alloc_round = headerBlock->default_entry_stride;
+		/* initialize common entry details.  */
+		entrytemp.detail.valid = 0;
+		entrytemp.detail.timestamped = 0;
+		entrytemp.detail.allocated = 1;
+		entrytemp.detail.__reserved = 0;
+		entrytemp.detail.category = 0;
+		entrytemp.detail.subcat = 0;
+		entrytemp.detail.len = (alloc_round / DEFAULT_ALLOC_UNIT);
+
+	} else {
+		debug_printf("%s(%p, %ld) type check failed\n", __FUNCTION__,queue, alloc_round);
+	}
+	return (entrytemp.idUnit);
+}
+
+static inline SPHLFEntryHeader_t *
+SPHMPMCQAdvanceHead(SPHLFEntryHeader_t **head_p, sphLFEntryID_t idAlloc, sphLFEntryID_t idFree, unsigned short len, longPtr_t qlo, longPtr_t qhi) {
+	SPHLFEntryHeader_t *entryPtr = *head_p;
+	if (entryPtr->entryID.idUnit == idFree) {
+		longPtr_t new_head = ((longPtr_t)entryPtr) + len;
+		if (new_head >= qhi)
+			new_head = qlo;
+		entryPtr->entryID.idUnit = idAlloc;
+		*(longPtr_t *)head_p = new_head;
+		return entryPtr;
+	}
+	return 0;
+}
+
+SPHLFEntryDirect_t
+SPHMPMCQAllocStrideDirectTM (SPHMPMCQ_t queue)
+{
+	volatile SPHMPMCQHeader *headerBlock = (SPHMPMCQHeader *) queue;
+	SPHLFEntryHeader_t *entryPtr = 0;
+
+	debug_printf("%-6d: %s NEW\n",sphdeGetTID(),__FUNCTION__);
+	if (SOMSASCheckBlockSigAndType ((SASBlockHeader *) headerBlock,SAS_RUNTIME_PCQUEUE_TM)) {
+		sphLFEntry_t entrytemp;
+		const sphLFEntry_t entryfree = {0};
+		const unsigned short stride = headerBlock->default_entry_stride;
+		const longPtr_t qlo = headerBlock->startq;
+		const longPtr_t qhi = headerBlock->endq;
+#ifdef __HTM__
+		TM_buff_type TM_buff;
+#endif
+
+		entrytemp.detail.valid = 0;
+		entrytemp.detail.timestamped = 0;
+		entrytemp.detail.allocated = 1;
+		entrytemp.detail.__reserved = 0;
+		entrytemp.detail.category = 0;
+		entrytemp.detail.subcat = 0;
+		entrytemp.detail.len = (stride / DEFAULT_ALLOC_UNIT);
+
+#ifdef __HTM__
```

Comment:
OK. Right now, if compiler does not support -mhtm and if --enable-htm is passed with configure, it fails during make phase.  It would be better to make it fail in the configure phase itself, but that can be done as a separate patch. We can continue with this for now.